tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/graph_partition.py

```diff
@@ -0,0 +1,518 @@
+  remote op layer to the outputs of the previous subgraph layer. We can
+  then traverse and construct the previous subgraph layer.
+
+  Each subgraph layer can be captured into one ExecutionSpec, but each
+  remote op layer need to be stored into one or more ExecutionSpecs based on
+  the number of remote ops. This happens because we assume that the execution
+  of ExecutionSpecs is strictly sequential.
+
+  Args:
+    graph_def: A `GraphDef` proto.
+    graph_output_names: A list of graph output node names.
+    graph: A tf.Graph representing the same graph as graph_def.
+    node_name_to_node_def: A mapping from node names to `NodeDef` protos.
+    remote_op_to_parents: A mapping from remote op name to a list of
+                          immediate remote op parent names.
+
+  Returns:
+    A list of ExecutionSpecs, where the order of the list represents the
+    order of execution.
+  """
+  execution_specs = []  # type: List[ExecutionSpec]
+  previously_visited = set()  # type: Set[Text]
+
+  for remote_op_layer in _RemoteOpLayers(remote_op_to_parents):
+    # Get one subgraph layer
+    output_node_names = _get_subgraph_layer_output_node_names(
+        remote_op_layer, node_name_to_node_def)
+
+    if output_node_names:
+      spec = _partition_one_subgraph_layer(
+          graph_def, graph, node_name_to_node_def,
+          previously_visited, output_node_names)
+      execution_specs.append(spec)
+      previously_visited |= set(_get_non_input_names(spec.subgraph))
+
+    # Get one remote op layer
+    specs = _partition_one_remote_op_layer(
+        remote_op_layer, node_name_to_node_def)
+    execution_specs.extend(specs)
+
+  # Get the last subgraph layer
+  spec = _partition_one_subgraph_layer(
+      graph_def, graph, node_name_to_node_def,
+      previously_visited, graph_output_names)
+  execution_specs.append(spec)
+
+  return execution_specs
+
+
+def _get_subgraph_layer_output_node_names(
+    remote_op_layer: List[Text],
+    node_name_to_node_def: Mapping[Text, node_def_pb2.NodeDef]
+) -> List[Text]:
+  """Gets the output node names of a subgraph layer.
+
+  We derive the output node names from the input names of the succeeding
+  remote op layer.
+
+  Args:
+    remote_op_layer: A list of remote op names for a remote op layer.
+    node_name_to_node_def: A mapping from node names to `NodeDef` protos.
+
+  Returns:
+    A list of output node names of a subgraph layer.
+  """
+  output_node_names = []
+
+  for remote_op_name in remote_op_layer:
+    for input_node_name in node_name_to_node_def[remote_op_name].input:
+      input_node = node_name_to_node_def[input_node_name]
+
+      if input_node_name in output_node_names:
+        continue
+      # Assumption: Later in beam, the graph inputs will be loaded first.
+      if _is_placeholder_op(input_node):
+        continue
+
+      output_node_names.append(input_node_name)
+
+  return output_node_names
+
+
+def _partition_one_subgraph_layer(
```

Comment:
Optional: name this `_parent_subgraph_layer_to_execution_spec` or `_get_execution_spec_for_parent_subgraph_layer` or something more appropriate (it doens't actually do any partitioning.

Reply:
Addressed.